File: python/day3.py
import re

def get_input():
    with open('../inputs/day3.txt', 'r', encoding='utf8') as file:
        data = file.read()
    return data

def process_input_a(input):
    regex_matches = re.findall(r'mul\(\d{1,3},\d{1,3}\)', input)
    functions = [str(match) for match in regex_matches]
    return functions

def parse_multipliers_a(functions):
    multipliers = []
    for func in functions:
        a, b = map(int, re.findall(r'\d{1,3}', func))
        multipliers.append([a, b])
    return multipliers

def process_input_b(input):
    regex_matches = re.findall(r'mul\(\d{1,3},\d{1,3}\)|do\(\)|don\'t\(\)', input)
    functions = [str(match) for match in regex_matches]
    return functions

def parse_multipliers_b(functions):
    multipliers = []
    is_enabled = True
    for func in functions:
        if func == "do()":
            is_enabled = True
        elif func == "don't()":
            is_enabled = False
        elif "mul" in func and is_enabled:
            a, b = map(int, re.findall(r'\d{1,3}', func))
            multipliers.append([a, b])
    return multipliers

def solution(reports):
    return sum(a * b for a, b in reports)


input_data = get_input()
functions_a = process_input_a(input_data)
multipliers_a = parse_multipliers_a(functions_a)
print(solution(multipliers_a))
functions_b = process_input_b(input_data)
multipliers_b = parse_multipliers_b(functions_b)
print(solution(multipliers_b))

File: typescript/day3.ts
import fs from 'fs';

function getInput():string{
    const data = fs.readFileSync('../../inputs/day3.txt', 'utf8');

    return data;
}

function processInputA(input: string): Array<string>{
    const regexMatches = input.match(/mul\(\d{1,3},\d{1,3}\)/g)?.map(String)
    let functions:Array<string> = []
    if(regexMatches)
        functions = regexMatches.map(String)
    
    return functions
}

function parseMultipliersA(functions:Array<string>):Array<Array<number>>{
    let multipliers:Array<Array<number>> = []
    for(let func of functions){
        const [a, b] = func.match(/\d{1,3}/g)?.map(Number) as Array<number>
        multipliers.push([a, b])
    }

    return multipliers

}

function processInputB(input: string): Array<string>{
    const regexMatches = input.match(/mul\(\d{1,3},\d{1,3}\)|do\(\)|don't\(\)/g)?.map(String)
    let functions:Array<string> = []
    if(regexMatches)
        functions = regexMatches.map(String)
    
    return functions
}

function parseMultipliersB(functions:Array<string>):Array<Array<number>>{
    let multipliers:Array<Array<number>> = []
    let isEnabled = true
    for(let i = 0; i < functions.length; i++){
       if(functions[i] === "do()"){
           isEnabled = true
       }
       if(functions[i] === "don't()"){
           isEnabled = false
       }
       if(functions[i].includes("mul")){
           if(isEnabled){
               const [a, b] = functions[i].match(/\d{1,3}/g)?.map(Number) as Array<number>
               multipliers.push([a, b])
           }
       }
    }

    return multipliers
}

function solution(reports:Array<Array<number>>):number{
    let sum = 0

    for(let [a, b] of reports){
        sum += a * b
    }
    return sum;
    
}


function main(){
    const input = getInput();
    const functionsA = processInputA(input)
    const multipliersA = parseMultipliersA(functionsA)
    console.log(solution(multipliersA))
    const functionsB = processInputB(input)
    const multipliersB = parseMultipliersB(functionsB)
    console.log(solution(multipliersB))
}

main();